Simplify dockerfile, use inline commands for containers, CA server deployment, security configs, readme updates, request limits, sts config

pepov opened the pull request:
- update Zookeeper dependency\
 - update Dockerfile to avoid unnecessary environment variables and to use the nifi user
 - update Loadbalancer to use port 80
 - resource requests/limits for log containers

pepov committed:
Add docs and tune chart to run smooth in minikube and on GKE. Also:
 - update Zookeeper dependency\
 - update Dockerfile to avoid unnecessary environment variables and to use the nifi user
 - update Loadbalancer to use port 80
 - resource requests/limits for log containers

pepov committed:
CA server deployment (initial)

markap14 commented:
Hey @pepov when I tried this, I got the following error:

[nifi@nifi-04 apache-nifi]$ helm upgrade --install apache-nifi .
Release "apache-nifi" does not exist. Installing it now.
Error: apiVersion "apps/v1" in apache-nifi/templates/ca.yaml is not available

pepov commented:
Seems to me that apps/v1 apiversion for Deployment was only introduced in 1.9. What version of kubernetes do you use? GKE uses 1.8 by default, but 1.9 is already available. Minikube is already on 1.10

pepov committed:
README updates

pepov committed:
secure bootstrap by default using generated client certificates. opt-out to go with unsecured cluster

pepov committed:
remove custom entrypoint from Dockerfile, control everything from the chart

pepov committed:
restore replicaCount 2 to be the default, also use Parallel as default podManagementPolicy

markap14 commented:
I'm using 1.8. I feel like we should probably try to avoid using something that is not available in 1.8 if that's the default for GKE. Is that possible?

pepov commented:
Yeah, sure, sorry for that, I'm changing it then to v1beta2 which will be supported for some time.

pepov committed:
Change deployment apiversion to be compatible with 1.8 clusters

markap14 commented on docker/Dockerfile:
```diff
@@ -53,27 +53,19 @@ RUN curl -fSL ${MIRROR}/${NIFI_BINARY_URL} -o ${NIFI_BASE_DIR}/nifi-${NIFI_VERSI
+
 COPY ./files/nifi-env.sh $NIFI_HOME/bin/nifi-env.sh
 
 COPY ./files/docker-entrypoint.sh /
```
Believe we need to remove this line as well.

pepov replied:
yes and I will do a diff with the official one to see how much we diverged

markap14 commented on values-gcp.yaml:
```diff
@@ -0,0 +1,3 @@
+storageClass: nifi-storage-class
```
This is a much smarter way to handle this, to put it into its own values-gcp.yaml file. Didn't realize we could use multiple -f arguments with helm. Probably makes sense to rename the storageClass to something like "gce-ssd" though?

pepov replied:
Maybe I don't fully understand the question but I don't think we should rename it, since nifi-storage-class is an abstraction over more specific storage implementations.

markap14 replied:
that was originally my intent with the naming, but if it's being pulled out into a file called "values-gcp.yaml" then that seems weird to me... perhaps we should just rename it to values-pvc.yaml or something? It seems odd to name it 'nifi-storage-class' with the intent of that being generic and then put into a values file named after google cloud platform

pepov replied:
Maybe we could use another variable like `customStorageClass: true/false` and then we can name the storageClass whatever we wan't?

pepov replied:
I've just pushed it

markap14 commented on templates/statefulset.yaml:
```diff
@@ -29,11 +32,72 @@ spec:
+          cat "${NIFI_HOME}/conf/nifi.temp" > "${NIFI_HOME}/conf/nifi.properties"
+
+          prop_replace nifi.web.http.host ${FQDN}
+          prop_replace nifi.web.https.host ${FQDN}
```
We cannot set both nifi.web.http.host and nifi.web.https.host - we must set only https.host if we expect user to login via https. Otherwise, we must set only http.host. If we set both, nifi redirects to the login page on http, which causes and error. In the next version of NiFi, the server won't even start if both of these properties are configured. So we can probably just set nifi.web.https.host to ${FQDN} below, in the "if .Values.ca.enabled" section and also there set nifi.web.http.host to empty string.

markap14 commented on templates/statefulset.yaml:
```diff
@@ -29,11 +32,72 @@ spec:
+          prop_replace nifi.security.keyPasswd $(jq -r .keyPassword ${NIFI_BASE_DIR}/data/cert/config.json)
+
+          prop_replace nifi.security.truststore ${NIFI_BASE_DIR}/data/cert/truststore.jks
+          prop_replace nifi.security.truststorePasswd $(jq -r .trustStorePassword ${NIFI_BASE_DIR}/data/cert/config.json)
```
In this case, we also need to set the "nifi.web.proxy.host" to the IP address and/or FQDN of the load balancer. Otherwise, we can't connect to a secured nifi instance.

markap14 commented on values.yaml:
```diff
@@ -111,29 +124,13 @@ zookeeper:
+ca:
+  enabled: true
+  port: 8443
+  token: secret-token-to-prevent-mitm
```
I don't know how this would work exactly, but I am wondering if we could make this value 'optional' and if not specified, generate a UUID or something like that? Does Helm provide the ability to generate a unique ID?

pepov replied:
We can generate it using a script and upload it as a kubernetes secret, or I can do the same inside a pod. Will do it separately.

pepov committed:
Proper handling of http/https ports, remove obsolete ingress, disable loadbalancer for minikube, but enable hostPort 443 and the webProxyHost setting

pepov committed:
admin cert generator pod with cert download support, to make it easier to access a secured cluster with client key and certificate

pepov committed:
Fixing issues around proper/graceful shutdown

pepov committed:
customStorageClass flag to replace nifi-storage-class

pepov committed:
implement liveness/readyness probes

pepov committed:
controlled alpine-tini image

pepov committed:
fix shebang and remove failing on errors to avoid missing the error message

pepov committed:
Note on the unathorized proxy issue when adding a new node above the initial cluster size, need to deal with it programatically

pepov committed:
Set needClientAuth flag together with other secure properties

pepov committed:
Add init container to delay starting up until zookeeper is ready

pepov committed:
Fix node cert generator to include the explicit DN instead of relying on the default (which caused issues on GKE)

pepov committed:
Disable Parallel podManagementPolicy as it caused issues during startup

pepov committed:
separate helm values file for minimal zookeeper resource setup

pepov committed:
Use PROXY_HOST env to set the domain name of the proxy, that the cluster will be accessed through

pepov committed:
Change authorizations.xml location to persist between restarts

pepov committed:
Use release name when checking for the zookeeper service dependency

pepov committed:
Tune liveness/readiness check params and add graceful shutdown script

pepov committed:
Install script, admin cert script fixes, added useful notes

pepov committed:
use pepov/nifi-node temporarily

pepov committed:
Tolerate unready state in the headless service so that additional nodes could join the cluster

pepov committed:
Add default hard AntiAffinity

pepov committed:
install and admincert script improvements

pepov committed:
custom admin ca

pepov committed:
Use nifi-toolkit image for CA and cert generation

pepov committed:
Use different authorizers.xml in case of initial and extended cluster configurations

pepov committed:
Use image built from https://github.com/apache/nifi/pull/2747 where log directory is in its original place. Switch back to Parallel podManagementPolicy for default

pepov committed:
Script refactor:
 - Rename env to profile
 - Require profile and context specified as arguments to avoid suprises with env vars
 - Move install.sh to scripts directory
 - Make sure no relative file paths are used
 - Arg checking moved to functions.sh for reusability
 - Fix helm init

pepov committed:
Create mitm token as kubernetes secret and mount it instead of passing it az a helm variable

pepov committed:
Destroy script

pepov committed:
Remove solved todos

pepov committed:
Add release label to secrets, remove on destroy by label

pepov committed:
Check faster for readyness

pepov committed:
Separate image tag from nifi version

pepov committed:
in-cluster ca-token-generator initials

pepov committed:
script fixes, ignore files, default replica count 1 for minikube

pepov committed:
in-cluster CA token generator

pepov committed:
add missing force flag to kubectl replace

pepov committed:
ca-cert-generator docker image and script to launch inside kubernetes

pepov committed:
generate secret in helm pre-install

pepov committed:
Move ca cert generator to run in kubernetes and save cert into secrets

pepov committed:
update helmignore

pepov committed:
Update README.md

pepov committed:
Update README.md

derfloscher commented:
First of all thank you for your effort on continuing development. Will this be the basis for the official hortonworks k8s nifi setup? I noticed you linking to their space, the repository however does not seem to exist. Can you already give a timeline on that? Cheers

pepov commented:
Hi @derfloscher yes I accidentally linked to it previously, however it's still a work in progress internally. You can expect something this year, but can't tell about the exact date.
